saleor/payment/utils.py

```diff
@@ -0,0 +1,180 @@
+    return payment
+
+
+def create_transaction(
+        payment: Payment,
+        token: str,
+        transaction_type: str,
+        is_success: bool,
+        amount: Decimal,
+        currency: str,
+        gateway_response: Optional[Dict] = None) -> Transaction:
+    if not gateway_response:
+        gateway_response = {}
+    txn, _ = Transaction.objects.get_or_create(
+        payment=payment,
+        token=token,
+        transaction_type=transaction_type,
+        is_success=is_success,
+        amount=amount,
+        currency=currency,
+        gateway_response=gateway_response)
+    return txn
+
+
+def gateway_get_transaction_token(provider_name: str):
+    # FIXME Add tests
+    provider, provider_params = get_provider(provider_name)
+    return provider.get_transaction_token(**provider_params)
+
+
+@validate_payment
+def gateway_authorize(payment: Payment, transaction_token: str) -> Transaction:
+    if not payment.charge_status == ChargeStatus.NOT_CHARGED:
+        raise PaymentError('Charged transactions cannot be authorized again.')
+
+    provider, provider_params = get_provider(payment.variant)
+    with transaction.atomic():
+        txn, error = provider.authorize(
+            payment, transaction_token, **provider_params)
+    # FIXME Create an order event ?
+    if not txn.is_success:
+        raise PaymentError(error)
+    return txn
+
+
+@validate_payment
+def gateway_capture(payment: Payment, amount: Decimal) -> Transaction:
+    if amount <= 0:
+        raise PaymentError('Amount should be a positive number.')
+    if payment.charge_status not in {ChargeStatus.CHARGED,
+                                     ChargeStatus.NOT_CHARGED}:
+        raise PaymentError('This payment cannot be captured.')
+    if amount > payment.total or amount > (
+            payment.total - payment.captured_amount):
+        raise PaymentError('Unable to capture more than authorized amount.')
+
+    provider, provider_params = get_provider(payment.variant)
+    with transaction.atomic():
+        txn, error = provider.capture(
+            payment, amount=amount, **provider_params)
+        if txn.is_success:
+            payment.charge_status = ChargeStatus.CHARGED
+            payment.captured_amount += txn.amount
+            payment.save(update_fields=['charge_status', 'captured_amount'])
+            order = payment.order
+            if order and order.is_fully_paid():
+                handle_fully_paid_order(order)
+            # FIXME Create an order event ?
+    if not txn.is_success:
+        raise PaymentError(error)
+    return txn
+
+
+@validate_payment
+def gateway_void(payment) -> Transaction:
+    if not can_be_voided(payment):
+        raise PaymentError('Only pre-authorized transactions can be voided.')
+    provider, provider_params = get_provider(payment.variant)
+    with transaction.atomic():
+        txn, error = provider.void(payment, **provider_params)
```

Comment:
Wouldn't it be better to have a base behavior for non-mandatory features? Canceling an order is not supported by every provider, it would make sense to have a default behavior thanks to a `getattr`.

Reply:
I was thinking about `get_transaction_token` too.